pkg/vmconfig/vmconfig_reconciler.go

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package vmconfig
+
+import (
+	"context"
+
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/mo"
+	vimtypes "github.com/vmware/govmomi/vim25/types"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha3"
+)
+
+// Reconciler is a type that modifies a VirtualMachine's configuration before
+// and/or after creating or updating the underlying, platform resource.
+type Reconciler interface {
```

Comment:
I know you are trying to introduce a pattern for this.  But reading through this code and the diff in the PR, my opinion is that we introduce this pattern when we have a _need_ for it.  Right now, it seems a bit too complicated for something that might not even be used by other components (if we never get to breaking the current pre/post power on reconfigure thing in this mode).

If you don't mind, I would recommend shelving all this, keeping this change specific to reconciling the StorageClasses and the VMs that are referencing them.

Reply:
I appreciate the feedback, but I am going to keep this. It is part of my vision going forward, and this change highlights *that* the pattern works and *how* the pattern works. I initially did this change without the pattern, and it ended up looking very similar, just without the wrapper. All of VM Op should start adopting this pattern in order to make our reconciler less fragile than it is and so difficult to change.